### PR TITLE
Fix transform query for external database

### PR DESCRIPTION
**Pull request by ucasfl**
### Changelog category (leave one):

- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix and improve transform query for external database, we should recursively obtain all compatible predicates.

Found in case with predicate push down. for example, we have query like:

```sql
SELECT *
FROM
(
    SELECT *
    FROM t
    WHERE (a > 10) AND (toUInt64OrZero(timestamp_) > 12345)
)
WHERE c < 10
```
After predicate push down, the subquery become:
```sql
SELECT *
FROM t
WHERE ((a > 10) AND (toUInt64OrZero(timestamp_) > 12345)) and ((c < 10))
```
Before, the predicates `((a > 10) AND (toUInt64OrZero(timestamp_) > 12345))` will be dropped totally, only `c < 10` left, now, we will get predicates `(a > 10) and (c < 10)`.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 2d45213f70fc73c069d086519a7c49bf8cf6a861 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57888/2d45213f70fc73c069d086519a7c49bf8cf6a861/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7219851362">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57888/2d45213f70fc73c069d086519a7c49bf8cf6a861/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Storages/transformQueryForExternalDatabase.cpp [hunk 1]
```diff
@@ -329,14 +329,26 @@ String transformQueryForExternalDatabaseImpl(
         }
         else if (auto * function = original_where->as<ASTFunction>())
         {
-            if (function->name == "and")
+            if (function->name == "and" || function->name == "tuple")
```
Comment (vdimir) [inline src/Storages/transformQueryForExternalDatabase.cpp:332]: I don't really get where `tuple` comes from in your example. Could you please add a testcase?

**Review by vdimir — COMMENTED** [review]
Thank you for submitting PR!
Consider adding an example to test case (we do have some tests for mysql as well as unit-test `gtest_transform_query_for_external_database.cpp`)

File: src/Storages/transformQueryForExternalDatabase.cpp (hunk 1 above)
Comment (ucasfl) [reply src/Storages/transformQueryForExternalDatabase.cpp:332]: `((a > 10) AND (toUInt64OrZero(timestamp_) > 12345)) ` is a `tuple`.

**Review by vdimir — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**